### PR TITLE
[JENKINS-48113] slowing ajax call when request are slower.

### PR DESCRIPTION
**Pull request by aHenryJard**
See [JENKINS-48113](https://issues.jenkins-ci.org/browse/JENKINS-48113).

Given feedback on the other PR (https://github.com/jenkinsci/jenkins/pull/4994) trying another way (whitout js signature changes).
=> taking more time between 2 ajax calls when response time is long.
implementation here is to do 2 time longer when request response time is longer than time between 2 ajax calls.

### Proposed changelog entries

* Entry 1: Issue, Human-readable Text
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 1]
```diff
@@ -1658,9 +1658,25 @@ function expandTextArea(button,id) {
 // by using the contents fetched from the given URL.
 function refreshPart(id,url) {
     var intervalID = null;
+    var refreshInterval = 5000; //default is 5s
+
     var f = function() {
         if(isPageVisible()) {
             new Ajax.Request(url, {
+                onCreate: function() {
+                    requestStartTime = new Date().getTime();
```
Comment (mikecirioli) [inline war/src/main/webapp/scripts/hudson-behavior.js:1667]: Is this variable in scope so that it can be read on `onComplete` ?

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 2]
```diff
@@ -1658,9 +1658,25 @@ function expandTextArea(button,id) {
...
+                    console.log("onCreate send of id:"+id+", url:"+url+", start time:"+requestStartTime);
+                },
+                onComplete: function() {
+                    var requestDuration = new Date().getTime() - requestStartTime;
+                    if(requestDuration > refreshInterval) {
+                        //server or network has issues, slowing ajax calls. Will be reset by a page refresh (F5)
+                        refreshInterval = 2 * refreshInterval;
+                        window.clearInterval(intervalID);
+                        intervalID = window.setInterval(f, refreshInterval);
+                    }
```
Comment (mikecirioli) [inline war/src/main/webapp/scripts/hudson-behavior.js:1677]: need to add some logic so that the refresh interval can decrease if response time increases

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 1 above)
Comment (aHenryJard) [reply war/src/main/webapp/scripts/hudson-behavior.js:1667]: I tried it locally by changing the if it's working fine. I have no idea on how to slow answer to the Ajax calls, I don't know where this goes on server side.

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 2 above)
Comment (aHenryJard) [reply war/src/main/webapp/scripts/hudson-behavior.js:1677]: I don't think so, because a refresh or a page change reset to 5s so I think it's ok and less complex like that. If Jenkins is not responding people are going to hit F5.

Comment (mikecirioli) [reply war/src/main/webapp/scripts/hudson-behavior.js:1677]: People reloading the page seems like it would make this approach ineffective - possibly we could store the "delay" variable in a cookie.  This should also make it available across any page that makes ajax requests.

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 3]
```diff
@@ -1658,9 +1658,25 @@ function expandTextArea(button,id) {
...
 function refreshPart(id,url) {
     var intervalID = null;
+    var refreshInterval = 5000; //default is 5s
+
     var f = function() {
         if(isPageVisible()) {
             new Ajax.Request(url, {
+                onCreate: function() {
+                    requestStartTime = new Date().getTime();
+                    console.log("onCreate send of id:"+id+", url:"+url+", start time:"+requestStartTime);
```
Comment (MRamonLeon) [inline war/src/main/webapp/scripts/hudson-behavior.js:1668]:
```suggestion
```

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 4]
```diff
@@ -1658,9 +1658,25 @@ function expandTextArea(button,id) {
...
+                },
+                onComplete: function() {
+                    var requestDuration = new Date().getTime() - requestStartTime;
+                    if(requestDuration > refreshInterval) {
+                        //server or network has issues, slowing ajax calls. Will be reset by a page refresh (F5)
+                        refreshInterval = 2 * refreshInterval;
+                        window.clearInterval(intervalID);
+                        intervalID = window.setInterval(f, refreshInterval);
+                    }
+                    console.log("Complete send of id:"+id+", url:"+url+" -- duration is "+ requestDuration );
```
Comment (MRamonLeon) [inline war/src/main/webapp/scripts/hudson-behavior.js:1678]:
```suggestion
```

**Review by MRamonLeon — CHANGES_REQUESTED** [review]
I think the console should be avoided.

**Comment by oleg-nenashev** [comment]
@aHenryJard Hi, are you still interested in this pull request?

**Comment by aHenryJard** [comment]
I don't have time to go back to that, and it's too old in my mind now. I prefer to close it for now, as I cannot continue.
